src/websocket_client.rs

```diff
@@ -0,0 +1,171 @@
+            } else if req.uri().path().starts_with("/jet/create") {
+                let uuid = Uuid::new_v4();
+                if let Ok(mut jet_associations) = self.jet_associations.try_lock() {
+                    jet_associations.insert(uuid, None);
+                    *response.body_mut() = Body::from(uuid.to_string())
+                }
+            } else {
+                *response.status_mut() = StatusCode::BAD_REQUEST;
+            },
+
+            _ => {
+                *response.status_mut() = StatusCode::BAD_REQUEST;
+            }
+        }
+
+        Box::new(futures::future::ok::<Response<Body>, hyper::Error>(response))
+    }
+}
+
+
+fn process_req(req: &Request<Body>) -> Response<Body> {
+    /*
+        Source: https://gist.github.com/bluetech/192c74b9c4ae541747718ac4f4e20a14
+        Author: Ran Benita<bluetech> (ran234@gmail.com)
+    */
+
+    fn convert_key(input: &[u8]) -> String {
+        const WS_GUID: &[u8] = b"258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
+        let mut digest = sha1::Sha1::new();
+        digest.update(input);
+        digest.update(WS_GUID);
+        base64::encode(&digest.digest().bytes())
+    }
+    fn connection_has(value: &header::HeaderValue, needle: &str) -> bool {
+        if let Ok(v) = value.to_str() {
+            v.split(',').any(|s| s.trim().eq_ignore_ascii_case(needle))
+        } else {
+            false
+        }
+    }
+    let is_http_11 = req.version() == Version::HTTP_11;
+    let is_upgrade = req.headers()
+        .get(header::CONNECTION)
+        .map_or(false, |v| connection_has(v, "upgrade"));
+    let is_websocket_upgrade = req.headers()
+        .get(header::UPGRADE)
+        .and_then(|v| v.to_str().ok())
+        .map_or(false, |v| v.eq_ignore_ascii_case("websocket"));
+    let is_websocket_version_13 = req.headers()
+        .get(header::SEC_WEBSOCKET_VERSION)
+        .and_then(|v| v.to_str().ok())
+        .map_or(false, |v| v == "13");
+    if !is_http_11 || !is_upgrade || !is_websocket_upgrade || !is_websocket_version_13 {
+        return Response::builder()
+            .status(http::StatusCode::UPGRADE_REQUIRED)
+            .header(http::header::SEC_WEBSOCKET_VERSION, "13")
+            .body("Expected Upgrade to WebSocket version 13".into())
+            .unwrap();
+    }
+
+    let key = if let Some(value) = req.headers().get(header::SEC_WEBSOCKET_KEY) {
+        convert_key(value.as_bytes())
+    } else {
+        return Response::builder()
+            .status(StatusCode::BAD_REQUEST)
+            .body("".into())
+            .unwrap();
+    };
+
+    Response::builder()
+        .status(StatusCode::SWITCHING_PROTOCOLS)
+        .header(header::UPGRADE, "websocket")
+        .header(header::CONNECTION, "upgrade")
+        .header(header::SEC_WEBSOCKET_ACCEPT, key.as_str())
+        .header(header::SEC_WEBSOCKET_PROTOCOL, "binary")
+        .body(Body::empty())
+        .unwrap()
+}
+
+
+fn uuid_from_path(path: &str) -> Option<Uuid> {
```

Comment:
the way the URL is parsed and then deconstructed appears a bit flaky to me, maybe use regular expressions to catch specific patterns for /jet/create/<uuid>, /jet/connect/<uuid>, /jet/accept/uuid? this can be cleaned up later though.

Reply:
For sample usage of regex to parse URLs, you can take a look here: https://github.com/Devolutions/qui-vive-rs/blob/master/src/service.rs#L85